include/range/v3/view/shared.hpp

```diff
@@ -0,0 +1,154 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Filip Matzner 2017
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+#ifndef RANGES_V3_VIEW_SHARED_HPP
+#define RANGES_V3_VIEW_SHARED_HPP
+
+#include <memory>
+#include <type_traits>
+#include <meta/meta.hpp>
+#include <range/v3/begin_end.hpp>
+#include <range/v3/iterator_range.hpp>
+#include <range/v3/range_concepts.hpp>
+#include <range/v3/size.hpp>
+#include <range/v3/view/all.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        template<typename T>
+        struct shared_view
+          : view_interface<shared_view<T>>
+        {
+        private:
+            // shared storage
+            std::shared_ptr<T> rng_ptr_;
+
+            // range of the shared storage
+            using range_t = view::all_t<T>;
+            range_t rng_;
```

Comment:
I also wanted to call the stored range directly, but I was getting SEGFAULTs in tests. It was a bug in view::cycle though (#595), so I will do it as you say.